akka-stream-tests/src/test/scala/akka/stream/io/FramingSpec.scala

```diff
@@ -0,0 +1,182 @@
+import akka.stream.scaladsl.{ Sink, Source, Flow }
+import akka.stream.stage.{ TerminationDirective, SyncDirective, Context, PushPullStage }
+import akka.stream.testkit.AkkaSpec
+import akka.util.{ ByteString, ByteStringBuilder }
+
+import scala.collection.immutable
+import scala.concurrent.Await
+import scala.concurrent.duration._
+import scala.concurrent.forkjoin.ThreadLocalRandom
+
+class FramingSpec extends AkkaSpec {
+
+  val settings = ActorFlowMaterializerSettings(system)
+  implicit val materializer = ActorFlowMaterializer(settings)
+
+  class Rechunker extends PushPullStage[ByteString, ByteString] {
+    private var rechunkBuffer = ByteString.empty
+
+    override def onPush(chunk: ByteString, ctx: Context[ByteString]): SyncDirective = {
+      rechunkBuffer ++= chunk
+      rechunk(ctx)
+    }
+
+    override def onPull(ctx: Context[ByteString]): SyncDirective = {
+      rechunk(ctx)
+    }
+
+    override def onUpstreamFinish(ctx: Context[ByteString]): TerminationDirective = {
+      if (rechunkBuffer.isEmpty) ctx.finish()
+      else ctx.absorbTermination()
+    }
+
+    private def rechunk(ctx: Context[ByteString]): SyncDirective = {
+      if (!ctx.isFinishing && ThreadLocalRandom.current().nextBoolean()) ctx.pull()
+      else {
+        val nextChunkSize =
+          if (rechunkBuffer.isEmpty) 0
+          else ThreadLocalRandom.current().nextInt(0, rechunkBuffer.size + 1)
+        val newChunk = rechunkBuffer.take(nextChunkSize)
+        rechunkBuffer = rechunkBuffer.drop(nextChunkSize)
+        if (ctx.isFinishing && rechunkBuffer.isEmpty) ctx.pushAndFinish(newChunk)
+        else ctx.push(newChunk)
+      }
+    }
+  }
+
+  val rechunk = Flow[ByteString].transform(() ⇒ new Rechunker).named("rechunker")
+
+  "Delimiter bytes based framing" must {
+
+    val delimiterBytes = List("\n", "\r\n", "FOO").map(ByteString(_))
+    val baseTestSequences = List("", "foo", "hello world").map(ByteString(_))
+
+    def completeTestSequences(delimiter: ByteString): immutable.Iterable[ByteString] =
+      for (prefix ← 0 until delimiter.size; s ← baseTestSequences)
+        yield delimiter.take(prefix) ++ s
+
+    "work with various delimiters and test sequences" in {
+      for (delimiter ← delimiterBytes; _ ← 1 to 100) {
+        val f = Source(completeTestSequences(delimiter))
+          .map(_ ++ delimiter)
+          .via(rechunk)
+          .via(Framing.delimiter(delimiter, 256))
+          .grouped(1000)
+          .runWith(Sink.head)
+
+        Await.result(f, 3.seconds) should be(completeTestSequences(delimiter))
+      }
+    }
+
+    "Respect maximum line settings" in {
+      // The buffer will contain more than 1 bytes, but the individual frames are less
+      Await.result(
+        Source.single(ByteString("a\nb\nc\nd\n")).via(Framing.lines("\n", 1)).grouped(100).runWith(Sink.head),
+        3.seconds) should ===(List("a", "b", "c", "d"))
+
+      an[IllegalStateException] should be thrownBy {
```

Comment:
The deciding question for me would be: “will this input always fail?” If yes, then it is an illegal argument, otherwise it depends on an illegal state.